### ISSUE TITLE
[RFC] Cross-Tokenizer On-Policy Distillation (OPD) Support

### ISSUE THREAD

**Issue by wsc16** [verl-project/verl-omni#382]
## Background: On-Policy Distillation (OPD)

OPD is a dense post-training paradigm where teacher per-token log-probabilities supervise student rollouts sampled from the student's own policy. Unlike off-policy SeqKD, OPD queries the teacher at states the student actually visits, eliminating exposure bias and propagating fine-grained distributional signal. Production recipes (Qwen3, MiMo, GLM-5, Thinking Machines Lab) all adopt OPD as the core post-training stage. verl-omni currently implements OPD for the diffusion line (`verl_omni/experimental/teacher_loop`, `verl_omni/workers/config/diffusion/distillation.py`) and for token-level distillation via the teacher runtime MVP merged in #325. And opd for qwen3Omni model is supported in #375.

The per-position objective (shared-tokenizer regime):

```
L_OPD = Σ_i D(q_i ‖ p_i)
```

where `q_i` is teacher next-token distribution over `V_T`, `p_i` is student distribution over `V_S`. This comparison is well-defined only when (a) `V_T = V_S`, and (b) index `i` denotes the same text prefix for both models.

## Current Tokenizer Constraint & Limitations

verl-omni's OPD implementation assumes the teacher and student share an identical tokenizer:

- Teacher loop (`verl_omni/experimental/teacher_loop/teacher_model.py`) scores student-generated sequences at the student's token coordinates, then the student consumes per-token advantages at those same coordinates.
- The PPO-style advantage `Â = log π_T(y_t) - log π_old(y_t)` is only meaningful when `y_t` resolves to the same text unit in both vocabularies.

This constraint restricts teacher selection to the student's model family. In practice, the strongest teacher for a target task often lives in a different family (e.g., Qwen3 vs GLM vs MiniMax vs DeepSeek), with vocab size, BPE merge table, and special-token handling all diverging. The span Jaccard between segmentations can drop to ~0.81 (BPM paper, MiniMax vs Qwen3.5 pair), at which point raw logprob transfer becomes semantically broken.

## Why Cross-Tokenizer OPD Is Necessary

Three converging motivations:

1. **Teacher heterogeneity drives capability integration.** Consolidating complementary strengths (Qwen3 reasoning, GLM-Z1 coding, MiniMax agentic) into one compact student is the dominant deployment pattern. None of these share tokenizers.
2. **Compute efficiency over SeqKD.** Cross-tokenizer OPD has been shown to match or exceed off-policy SeqKD at a fraction of the rollout cost (DPCA paper, Table 1). verl-omni currently has no path to this.
3. **Existing fallbacks are weak.** When tokenizers differ today, users fall back to SFT on teacher-generated responses — discarding the teacher's full distributional supervision entirely.

## Survey: Cross-Tokenizer OPD Methods

Four representative methods have emerged in 2026. All operate on the same premise: although teacher/student tokenizers differ, both decode the same underlying byte/text stream, so an alignment bridge can be constructed.

### Common formulation

Both models produce token sequences over the same response text `x`:

- Teacher: `T(x) = (w^t_1, ..., w^t_M)`, vocab `V_T`
- Student: `S(x) = (w^s_1, ..., w^s_N)`, vocab `V_S`

The challenge is producing a per-position teacher target that lives on `V_S` (or a comparable common space), so the student OPD loss can consume it.

### Method 1: BPM — Byte-Prefix Marginalization

*Wang et al., "Cross-Tokenizer OPD via Byte-Prefix Marginalization", arXiv:2607.22334v1 (2026-07)*

**Idea.** Marginalize the teacher next-token distribution over the student vocabulary in a shared byte space.

**Algorithm.**

1. Build a byte-trie of `V_S` once per tokenizer pair.
2. Define `φ(w) = argmax_{u ∈ V_S, bytes(u) ⊆ bytes(w)} |bytes(u)|` — the longest student token whose byte representation prefixes the teacher token's bytes. `φ(w)=⊥` if none.
3. Target over `V_S`:
   ```
   t_i(u) = Σ_{w: φ(w)=u} q_{j(i)}(w)
   t_i(⊥) = Σ_{w: φ(w)=⊥} q_{j(i)}(w)
   ```
4. Spanning-token correction (student token spans multiple teacher tokens, < 1% of positions): use the chain-rule lower bound `t_i(u) = Π_k q_k(w^t_k) · μ_{j+m}(r)`.
5. Stop-token bridge target trained under D_0 (forward KL) to recover from stopping collapse.

**Loss.** Generalized JSD: `D_α(t‖p) = α·KL(t‖m) + (1-α)·KL(p‖m)`, with α=0 (forward KL) winning empirically.

**Properties.** Satisfies three desiderata by construction:

- D1 Vocabulary Completeness: target defined for every `u ∈ V_S ∪ {⊥}`
- D2 Byte-level Alignment: mapping keyed on byte representation
- D3 Mass Preservation: every teacher token routed to exactly one cell

**Strength.** Theoretical exactness: recovers the teacher-induced byte-prefix marginal at > 99% of training positions. Lowest information loss among the four.

**Cost.** Pre-compile trie per tokenizer pair; per-step scatter of teacher logits onto `V_S`.

### Method 2: SimCT — Minimal Aligned Units

*Sun et al., "SimCT: Recovering Lost Supervision for Cross-Tokenizer OPD", arXiv:2605.07711v2 (2026-05). Code: https://github.com/sunjie279/SimCT-*

**Idea.** Do not modify the OPD loss. Expand the supervision space `U` to include "minimal aligned units" (MAUs) — the finest jointly tokenizable text continuations.

**Algorithm.**

1. `U_SimCT = (V_T ∩ V_S) ∪ A`, where `A` is the set of MAUs (texts both tokenizers can express, even if via different token boundaries).
2. For each `u ∈ U_SimCT`, score under model M:
   - 1:1 shared token: `s_M(u|x_<t) = log p_M(u|x_<t)`
   - MAU with M-realization `(v_1,...,v_k)`: `s_M(u|x_<t) = (1/k) Σ_j log p_M(v_j|x_<t, v_<j)` — length-normalized continuation likelihood.
3. Normalize: `q_M(u|x_<t) ∝ exp(s_M(u|x_<t))` over a finite candidate set.
4. Standard reverse-KL OPD loss between `q_S` and `q_T` on `U_SimCT`.

**Properties.**

- Not mass-preserving: it is an operational scoring interface, not a generative projection.
- Theorem 1: MAUs are the finest boundary-consistent interface jointly expressible by both tokenizers.
- Proposition 1: coarsening removes within-unit teacher-student distinctions (KL gap = lost signal).

**Strength.** Minimal engineering change — loss form unchanged, only candidate set construction differs. Low wall-clock overhead vs SimpleOPD.

**Cost.** Candidate set enumeration per position; requires teacher-side scoring of MAU realizations.

### Method 3: DPCA — Dual-Pointer Chunk Alignment

*Niu et al., "Breaking the Tokenizer Barrier: On-Policy Distillation across Model Families", arXiv:2606.09456v1 (2026-06). Code: https://github.com/ivanniu/On-Policy-Distill*

**Idea.** Two-stage: (a) align teacher/student token sequences into minimal synchronized chunks via dual-pointer scan; (b) closed-form credit assignment within each chunk that preserves the student's internal semantic prior.

**Algorithm — Stage 1 (DPCA alignment).**

- Synchronization point `(i,j)`: `D(y^S_{1:i}) = D(y^T_{1:j})` (decoded prefixes equal).
- Greedy catch-up: maintain pointers `(p_S, p_T)`, extend the side with shorter decoded string until equal, emit chunk, advance.
- Theorem 4.2: completeness (finds all sync points) + minimality (no chunk further divisible), leveraging strict monotonicity of detokenization.

**Algorithm — Stage 2 (Credit assignment).**

- Within chunk `c`: student tokens `s_c = (s_1,...,s_m)`, teacher tokens `t_c`.
- Joint probability matching constraint: `Σ_i log π(s_{c,i}) = Σ_j log π_T(t_{c,j}) := L_{Tc}`.
- Optimization: distribute target log-probs preserving student semantic prior `p`:
  ```
  min_q (1/2) Σ_i (log q_i − k·log p_i)^2   s.t. Σ_i log q_i = L_{Tc}
  ```
- Lagrange closed form: `log q_i = (L_{Tc}/L_{Sc}) · log p_i`.
- Per-token advantage: `Â_{i,t} = (L_{Tc}/L_{Sc} − 1) · log p_i`. Reduces to `L_{Tc} − L_{Sc}` for 1:1 chunks (standard OPD).

**Loss.** PPO clip with cross-tokenizer advantage `Â`, directly pluggable into standard OPD trainer.

**Strength.** RL-native: advantage directly feeds PPO clip, zero trainer surgery. Student prior preserved within chunks.

**Cost.** Per-rollout dual-pointer scan; cheap closed-form solve. O(m+k) per chunk.

### Method 4: GOLD — Text-Aligned Rank Merging (Universal Logit Distillation + on-policy JSD)

*GOLD on-policy extension; closest published form: hybrid GOLD referenced in BPM/SimCT related work.*

**Idea.** Rank-based optimal-transport surrogate over the sorted probability vectors of teacher and student, plus a generalized JSD term over surface-matched tokens.

**Algorithm.**

1. Sort teacher and student next-token distributions by probability: `q^(1) ≥ q^(2) ≥ ...` and `p^(1) ≥ p^(2) ≥ ...`.
2. Match rank-by-rank: the loss minimizes a transport cost between the two sorted vectors, discarding *which* token each probability belongs to.
3. On-policy GOLD adds a generalized JSD term covering only tokens both tokenizers can express (surface-matched subset of `V_T ∩ V_S`).
4. For unmatched mass, rank pairing absorbs it into the sorted vectors.

**Properties.**

- Violates D2 (Byte-level Alignment): teacher mass may be assigned to student tokens with unrelated byte content, since rank pairing is content-agnostic.
- Hybrid: JSD on shared-vocab positions for content fidelity, sorted loss on the remainder for full-distribution coverage.

**Strength.** Vocab-agnostic — no byte/text alignment needed. Simple to implement given sorted top-k teacher logits.

**Cost.** Per-position sort over top-k teacher logits; transport computation. Lower fidelity than BPM/DPCA at the byte level.

## Comparative Analysis

| Dimension | BPM | SimCT | DPCA | GOLD |
|---|---|---|---|---|
| **Alignment granularity** | Byte prefix (student trie longest match) | Minimal aligned unit (jointly tokenizable text) | Minimal synchronized chunk (dual-pointer) | None (rank-based) |
| **Probability transform** | Exact byte-prefix marginal + chain lower bound for spanning | Length-normalized continuation score (not mass-preserving) | Chunk-ratio scaling, prior-proportional | Sorted-vector optimal transport |
| **Mass preserving** | ✅ (with explicit ⊥ residual) | ❌ (normalized scoring) | ✅ (chunk joint prob constraint) | Approximate (rank absorption) |
| **Vocab completeness** | ✅ (`V_S ∪ {⊥}`) | Partial (candidate set only) | ✅ (every student token gets advantage) | ✅ (full sort) |
| **Byte-level alignment** | ✅ | ❌ (text unit alignment) | ❌ (decoded-string-length alignment) | ❌ (rank only) |
| **OPD integration style** | Distribution target + generalized JSD | Distribution target + reverse KL | Per-token advantage + PPO clip | Sorted loss + JSD hybrid |
| **Theoretical exactness** | Exact at > 99% positions (Proposition C.2) | Finest jointly tokenizable interface (Thm 1) | Complete + minimal chunk partition (Thm 4.2) | Heuristic transport surrogate |
| **Failure modes** | Whitespace-position collapse (mitigated by precomputed mask) | Coarsening erases within-unit signal | Chunk-level ratio may miss intra-chunk distinctions | Identity lost: mass assigned to unrelated tokens |
| **Engineering complexity** | High (byte trie + scatter + bridge) | Low (candidate set + normalized scoring) | Medium (dual-pointer + closed-form) | Medium (sort + transport) |
| **RL-trainer fit** | Custom loss branch | Reuses OPD loss, swap candidate set | Native PPO advantage drop-in | Custom loss branch |
| **Best fit in verl-omni** | When byte-level fidelity matters most | Quick baseline, low surgery | PPO/actor-critic trainer drop-in | When teacher exposes only top-k logits |

### Summary by deployment scenario

- **verl PPO trainer already in place** → DPCA: advantage `Â` directly replaces same-tokenizer advantage in the existing clip loss. Lowest integration cost.
- **Theoretical rigor / dense supervision** → BPM: byte-trie + scatter is heavier but delivers the strictest D1-D3 guarantees. Best when teacher-student tokenizers are far apart (low span Jaccard).
- **Minimal change to existing OPD loop** → SimCT: keep reverse-KL loss, only swap the candidate set per position. Good first iteration.
- **Teacher exposes top-k logits only (no full-vocab)** → GOLD: rank-based transport works on top-k directly, no full distribution needed.

## Proposed Path for verl-omni

Suggested phased adoption:

1. **Phase 1 — DPCA path.** Implement dual-pointer alignment + closed-form advantage in the existing teacher loop (`verl_omni/experimental/teacher_loop/teacher_model.py`). PPO trainer (`verl_omni/trainer/`) consumes cross-tokenizer advantage unchanged. Lowest integration cost, unblocks cross-family teachers immediately.

2. **Phase 2 — BPM path.** Add byte-trie pre-compilation + teacher-logit scatter onto `V_S` for users needing strict D1-D3 guarantees. Expose as alternative advantage computation strategy in the teacher config (`verl_omni/workers/config/diffusion/distillation.py`).

3. **Phase 3 — SimCT / GOLD.** Add as lightweight baselines or for top-k-teacher-only settings.

### Open questions for maintainers

- Preferred alignment with existing diffusion distillation infrastructure (chunk alignment is text-domain; diffusion OPD operates in latent space — interface design needed).
- Whether to ship a unified `cross_tokenizer` config flag in `distillation.py` selecting among `bpm | simct | dpca | gold`.
- Handling for omni-modal models (Qwen3-Omni) where teacher/student vocab may include non-text modalities — byte alignment assumes text-only responses.

## References

- BPM: arXiv:2607.22334v1
- SimCT: arXiv:2605.07711v2 — https://github.com/sunjie279/SimCT-
- DPCA: arXiv:2606.09456v1 — https://github.com/ivanniu/On-Policy-Distill
- GOLD: hybrid on-policy universal logit distillation (referenced in BPM/SimCT related work; trl implementation tracks this line)
